optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,156 @@
+from optuna import logging
+from optuna.pruners.base import BasePruner
+from optuna.pruners.successive_halving import SuccessiveHalvingPruner
+from optuna import Study
+from optuna import type_checking
+
+if type_checking.TYPE_CHECKING:
+    from typing import List  # NOQA
+
+    from optuna.structs import FrozenTrial  # NOQA
+
+_logger = logging.get_logger(__name__)
+
+
+class HyperbandPruner(BasePruner):
+    """Pruner using Hyperband.
+
+    As SuccessiveHalving (SHA) requires the number of configurations
+    :math:`n` as its hyperparameter.  For a given finite budget :math:`B`,
+    all the configurations have the resources of :math:`B \\over n` on average.
+    As you can see, there will be a trade-off of :math:`B` and :math:`B \\over n`.
+    `Hyperband <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_ attacks this trade-off
+    by trying different :math:`n` values for a fixed budget.
+    Note that this implementation does not take as inputs the maximum amount of resource to
+    a single SHA noted as :math:`R` in the paper.
+
+    Args:
+        min_resource:
+            A parameter for specifying the minimum resource allocated to a trial noted as :math:`r`
+            in the paper.
+            See the details for :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        reduction_factor:
+            A parameter for specifying reduction factor of promotable trials noted as
+            :math:`\\eta` in the paper. See the details for
+            :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        min_early_stopping_rate_low:
+            The start point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+        min_early_stopping_rate_high:
+            The end point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            min_early_stopping_rate_low=0,
+            min_early_stopping_rate_high=4
```

Comment:
[MEMO] I think that we should consider what values are best as the defaults of the parameters before merging this PR.

Reply:
I think we should refer to other libraries' implementations.

Reply:
I agree with you.

Reply:
Inspired by Ray Tune, I'd like to use `brackets` or `num_brackets` instead of `min_early_stopping_rate_{low, high}` and want to make it defaults to 3 or 4 as the paper, Ray Tune, and Dask ML do so.

Reply:
[clarification] So, the default number of brackets will be changed to 3 or 4 from 5.

Reply:
Sounds good.
I think that we can run a benchmark to determine which is better, 3 or 4, as the default value.

Reply:
The above comment sounds like both `min_early_stopping_rate_low` and `min_early_stopping_rate_high` are removed, which means there's no way to change the `min_early_stopping_rate` of `SuccessiveHalvingPruner`. However, it's not desirable putting myself in the shoes of the compatibility with that pruner. Thus, I'd like to leave `min_early_stopping_rate_low` but with the order of the arguments changed as follows, assuming that most users do not change `min_early_stopping_rate`.

```python
class HyperbandPruner(BasePruner):

    ...

    def __init__(self, min_resource=1, reduction_factor=3, n_brackets=4, min_early_stopping_rate_low=0):
        ...
```

Reply:
### FYI: Benchmark Result

#### Figures

![hpo-bench-naval-880b007e3c61f4aecb7e9b0aa2f9be5fea9f491a076853f68f402769aa254034](https://user-images.githubusercontent.com/181413/71457527-a9e98600-27e1-11ea-8e18-6f8d1290d801.png)
![hpo-bench-parkinson-445bfa45fdbb8ec6ae6d4dba1909114f3948fa67b47209258db9291480b405b5](https://user-images.githubusercontent.com/181413/71457528-a9e98600-27e1-11ea-8d8d-25c56cae3f21.png)
![hpo-bench-protein-add73d4788d7900b34988a8b91cde43e820cac99f9e354e1e71b0ea0be3ef4a6](https://user-images.githubusercontent.com/181413/71457529-a9e98600-27e1-11ea-8dc2-3d4e34ab16cc.png)
![hpo-bench-slice-d88e1704447639bde17f236f7af47f93274d1f02bc8ec66733146ff9cdf50196](https://user-images.githubusercontent.com/181413/71457530-a9e98600-27e1-11ea-9252-b13474021e7a.png)

### Report

https://gist.github.com/sile/d9140bf45609a773f6edfbff94c5934b

Reply:
I think that `n_brackets=4`, which is the current implementation, is good as the default because it can cover a wider range of learning curve patterns than `n_brackets=3` and showed a  better benchmark result than `n_bracket=5`.